src/lib.rs

```diff
@@ -326,19 +323,18 @@ pub trait BatchDomain: Domain {
 }
 
 /// Trait that provides access to the components of an encrypted transaction output.
-///
-/// Implementations of this trait are required to define the length of their ciphertext
-/// field. In order to use the trial decryption APIs in this crate, the length must be
-/// either [`ENC_CIPHERTEXT_SIZE`] or [`COMPACT_NOTE_SIZE`].
-pub trait ShieldedOutput<D: Domain, const CIPHERTEXT_SIZE: usize> {
+pub trait ShieldedOutput<D: Domain> {
     /// Exposes the `ephemeral_key` field of the output.
     fn ephemeral_key(&self) -> EphemeralKeyBytes;
 
     /// Exposes the `cmu_bytes` or `cmx_bytes` field of the output.
     fn cmstar_bytes(&self) -> D::ExtractedCommitmentBytes;
 
-    /// Exposes the note ciphertext of the output.
-    fn enc_ciphertext(&self) -> &[u8; CIPHERTEXT_SIZE];
+    /// Exposes the note ciphertext of the output. Returns `None` if the output is compact.
+    fn enc_ciphertext(&self) -> Option<D::NoteCiphertextBytes>;
```

Comment:
This no longer "exposes" the output, and instead forcibly clones it, which is undesirable in usages of `ShieldedOutput` outside of `zcash_note_encryption`. We should instead return a reference that the caller can clone if desired (by adding a `Clone` bound on `Domain::NoteCiphertextBytes):
```suggestion
    /// Exposes the note ciphertext of the output.
    ///
    /// Returns `None` if the output is only compact.
    fn enc_ciphertext(&self) -> Option<&D::NoteCiphertextBytes>;
```

Also, document this change to the trait in the changelog.

Reply:
We tried to change the code to return references but encountered a few technical challenges due to how `NoteCiphertextBytes` and `CompactNoteCiphertextBytes` are implemented in both the `orchard` and `sapling-crypto` crates. These types are defined and implemented as newtypes over fixed-size byte arrays:

```rust
type NoteCiphertextBytes = NoteBytesData<{ ENC_CIPHERTEXT_SIZE }>;
type CompactNoteCiphertextBytes = NoteBytesData<{ COMPACT_NOTE_SIZE }>;

pub struct NoteBytesData<const N: usize>(pub [u8; N]);
```

The issue is that these newtypes wrap arrays that are not directly stored as such in the structs implementing `ShieldedOutput`. Therefore, we cannot return a reference to them directly because they are constructed on-the-fly from these arrays.

To enable returning references, we would need to adjust the trait to allow lifetimes and define associated types as generic with lifetimes using Rust's Generic Associated Types (GAT):

```rust
trait Domain {
    type NoteCiphertextBytes<'a>;
    type CompactNoteCiphertextBytes<'a>;
}
```

This approach would involve additional changes, impacting all `Domain` and `ShieldedOutput` implementations and adding complexity to the codebase. If using references is crucial, we can proceed with this strategy though.

Reply:
After discussion in the ZSA meeting, we will make my suggested change. If an action is non-compact, we know it is possible to get a reference to the appropriate type, and the data is large enough that we don't want to force clones.

Reply:
Done: now `enc_ciphertext` returns `Option` of a reference.